### PR TITLE
Relay modern

### PR DESCRIPTION
**Pull request by voxmatt**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by voxmatt** [comment]
Opening for reference -- not ready yet

**Comment by voxmatt** [comment]
@graphcool this is good to review now

@xtina-starr @alloy, I know you guys had also been working on some updated types. I'd love your eyes/comments on this.

**Comment by alloy** [comment]
@voxmatt Thanks, we’ll do so today or tomorrow :+1:

**Comment by aozgaa** [comment]
Waiting on review from @graphcool, @xtina-starr, and/or @alloy.

**Comment by aozgaa** [comment]
For the graphcool org, could one of @marktani, @schickling, or @sorenbs take a look? I'm unsure if you get notified when your org is mentioned. If not, consider adding yourselves as authors instead of the org.

**Comment by schickling** [comment]
@timsuchanek would be the best person to take a look!

**Comment by voxmatt** [comment]
Sorry to re-ping everyone, but if anyone has a chance to look at this, it would be greatly appreciated: @xtina-starr, @alloy, and @timsuchanek

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]
Commented on a couple things that require changes.

**Comment by alloy** [comment]
@voxmatt Pulling these into our app right now to do review, sorry for the delay.

File: types/relay-runtime/index.d.ts
Comment (alloy) [inline types/relay-runtime/index.d.ts]: Is it possible to easily expose more/all interfaces/types that Relay would otherwise do too, instead of having to jump through hoops like these?

E.g. I can import the `RecordSourceSelectorProxy` class from the `relay-runtime` module, but I cannot import its typings from this declaration file.

**Comment by alloy** [comment]
@voxmatt Nice (re last commit which exports all types) 👍 

Another thing I noticed is that you ported Flow’s `Maybe` types as a union of the type and `void`, but rather than `void` that should be `null | undefined`. I dunno how feasible it is to go through all of them, so alternatively I can send you a PR with locations I had to update while I was in the process of testing your types and PR the rest as we go (after merging this)?

**Comment by voxmatt** [comment]
I’d appreciate any help offered! I’m unsure how to structure the multiple packages inside of `react-relay`

**Comment by khell** [comment]
Great effort here. I started writing my own types as I needed them for Relay Modern and I'm glad to see our definitions are pretty close. :)

One difference is the definition of `ReactRelayQueryRenderer`, where I used generic types. This allows for type checking in the `render` prop function:
```javascript
    // ~~~~~~~~~~~~~~~~~~~~~
    // ReactRelayQueryRenderer
    // ~~~~~~~~~~~~~~~~~~~~~
    export interface QueryRendererProps<P> {
        cacheConfig?: RelayCommonTypes.CacheConfig;
        environment: RelayRuntimeTypes.Environment;
        query: GraphQLTaggedNode;
        render(readyState: ReadyState<P>): React.ReactElement<any> | undefined | null;
        variables: RelayCommonTypes.Variables;
        rerunParamExperimental?: RelayCommonTypes.RerunParam;
    }
    export interface ReadyState<P> {
        error: Error | undefined | null;
        props: P | undefined | null;
        retry?(): void;
    }
    export interface QueryRendererState<S> {
        readyState: ReadyState<S>;
    }
    export class ReactRelayQueryRenderer<P, S = undefined> extends React.Component<QueryRendererProps<P>, QueryRendererState<S>> { }
```

What do you think of this approach?

**Comment by voxmatt** [comment]
@khell this is great. Definitely better — I was mostly trying to power my way towards an MVP here. And sorry for the delay in progress, I've gotten bogged down at work.

**Comment by alloy** [comment]
@voxmatt How are you installing these typings into a project locally? I quickly tried to generate a package with `types-publisher` but got errors about unexpected relative paths being used (in imports).

**Comment by voxmatt** [comment]
@alloy they're actually broken currently. I was attempting to modify the patterns used to export the three packages inside `react-relay` to be more inline with what @uniqueiniquity suggested, but I stopped in the middle of figuring out the right pattern.

**Comment by alloy** [comment]
Oh I see, I had misunderstood [this comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19533#issuecomment-333581322). I’m gonna take a quick stab at getting them to work.

**Comment by andy-ms** [comment]
@alloy You shouldn't need `types-publisher` to test out new typings, just use the instructions from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#test).

File: types/react-relay/v1/index.d.ts
Comment (andy-ms) [inline types/react-relay/v1/index.d.ts]: If you're adding a newer version, that should go directly in `types/react-relay` and the old definitions should go in `types/react-relay/v0`, or be deleted if you probably won't be updating them.

File: types/relay-runtime/index.d.ts
Comment (andy-ms) [inline types/relay-runtime/index.d.ts]: Why does this need to be in a namespace? You could directly export types from the module and import them as in `import { RelayConcreteNode } from "relay-runtime";`.

File: types/relay-runtime/tsconfig.json
Comment (andy-ms) [inline types/relay-runtime/tsconfig.json]: Also add `"strictFunctionTypes": true,`

**Comment by alloy** [comment]
FYI I’ve done some work on top of @voxmatt’s work, which is on a WIP PR here https://github.com/voxmatt/DefinitelyTyped/pull/1. @andy-ms I’m gonna make that namespace change there as well.

File: types/relay-runtime/index.d.ts
Comment (alloy) [reply types/relay-runtime/index.d.ts]: Addressed in https://github.com/voxmatt/DefinitelyTyped/pull/2

File: types/relay-runtime/tsconfig.json
Comment (alloy) [reply types/relay-runtime/tsconfig.json]: @voxmatt This will require this branch to first be rebased (after merging https://github.com/voxmatt/DefinitelyTyped/pull/2 please 😄) and then add this to the tsconfig files.

**Comment by alloy** [comment]
I think that once https://github.com/voxmatt/DefinitelyTyped/pull/2 is merged and https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19533#discussion_r145273971 is addressed this PR will be in a good enough state to merge. @voxmatt will you address the latter or shall I create a new PR with a rebased version instead?

File: types/react-relay/v1/index.d.ts
Comment (alloy) [reply types/react-relay/v1/index.d.ts]: Got rid of those.

**Comment by voxmatt** [comment]
@alloy awesome; on it!

File: types/react-relay/v1/classic.d.ts
Comment (alloy) [inline types/react-relay/v1/classic.d.ts]: Did you re-add these v1 files on purpose? I don’t think we really need them, it’s not like Relay 0.x is going to get any further updates afaik.

File: types/relay-runtime/index.d.ts
Comment (voxmatt) [reply types/relay-runtime/index.d.ts]: I merged @alloy's PR  because it's just awesome to have the help and it was a big improvement. But I did prefer my pattern of only exporting the exports of the source package. The namespace enabled access to the underlying types without polluting the package's exports. Now, `import { ... } from 'react-relay'` exposes all sorts of cruft that makes it harder to understand the source package.

**Comment by alloy** [comment]
For other reviewers, this is a WIP PR where I use the new types in our codebase https://github.com/artsy/emission/pull/791.

File: types/react-relay/v1/classic.d.ts
Comment (voxmatt) [reply types/react-relay/v1/classic.d.ts]: @alloy sorry, just saw this _after_ merge. @uniqueiniquity had requested that I follow the patterns for updating major versions, which requires maintaining the previously existing types. https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19533#pullrequestreview-63426763

Comment (alloy) [reply types/react-relay/v1/classic.d.ts]: I see. OTOH @andy-ms notes that [we can delete them](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19533#discussion_r144883373) if if we’re not going to update them anymore, which is the case. People using e.g. Relay 0.10 can just pin to the corresponding `@types/react-relay` package.

Comment (alloy) [reply types/react-relay/v1/classic.d.ts]: Anyways, I’m fine with it if you wanna re-add them, but I think that’s the change that’s causing CI to fail atm, although its output is sparse to say the least.

Comment (voxmatt) [reply types/react-relay/v1/classic.d.ts]: Hey, if we can delete them, I'm all for it. @uniqueiniquity could you weigh in here?

Comment (alloy) [reply types/react-relay/v1/classic.d.ts]: Either way, shouldn’t this be `v0` instead of `v1`?

Comment (uniqueiniquity) [reply types/react-relay/v1/classic.d.ts]: Go ahead with @andy-ms's recommendation. If they won't be updated in the future, then it's fine to delete them.

**Comment by alloy** [comment]
My bad, my local setup was messed up. This should fix all linter issues https://github.com/voxmatt/DefinitelyTyped/pull/3.

File: types/relay-runtime/index.d.ts
Comment (alloy) [reply types/relay-runtime/index.d.ts]: I agree with this, but then we should just model the various modules (files) the same way they are in the distribution library. E.g. rather than having a `RelayNetworkTypes` section inside `types/relay-runtime/index.d.ts`, they should go into  `types/relay-runtime/lib/RelayNetworkTypes.d.ts`.

I haven’t taken a stab at this yet, because I wanted to reach consensus first and have limited time. If you agree with this approach then we should do so before merging, otherwise we’d introduce breaking changes next time.

Comment (andy-ms) [reply types/relay-runtime/index.d.ts]: It's pretty standard on DefinitelyTyped to bundle types for a library's interfaces along with types for its functions -- I would think it would be annoying to have to import from a different file just to get access to interfaces.

Comment (voxmatt) [reply types/relay-runtime/index.d.ts]: Totally fair. I'm in agreement with you two and, @alloy, thanks again for all the work getting this over the finish line. I took a weekend getting this in place and then have been overwhelmed with work ever since.

Comment (alloy) [reply types/relay-runtime/index.d.ts]: Ace 👌 I had the exact same situation, so luckily we could collab. High fives all around!

**Review by uniqueiniquity — APPROVED** [review]

**Comment by alloy** [comment]
🎉

**Comment by voxmatt** [comment]
Woot!

**Pull request merged**
